src/OpenLoco/Paint/PaintIndustry.cpp

```diff
@@ -0,0 +1,287 @@
+                        }
+                        adjustedBuildingPart += frameMask & tickThing;
+                    }
+                    else
+                    {
+                        if (!animationSequence.empty())
+                        {
+                            auto tickThing = (ticks >> buildingAnimation.animationSpeed) & (animationSequence.size() - 1);
+                            adjustedBuildingPart += animationSequence[tickThing];
+                        }
+                    }
+                    const auto sectionHeight = indObj->buildingPartHeight[adjustedBuildingPart];
+                    const uint32_t imageIdx = adjustedBuildingPart * 4 + indObj->var_12 + rotation;
+                    ImageId image = baseColour.withIndex(imageIdx);
+                    if (sectionCount == 0 && !baseColour.isBlended())
+                    {
+                        image = image.withNoiseMask((sectionProgress + 1) & 0x7);
+                    }
+                    session.addToPlotListAsChild(image, { 0, 0, height }, { -8, -8, bbZOffset }, { 38, 38, bbLengthZ });
+                    height += sectionHeight;
+                    sectionCount--;
+                }
+                if (totalSectionHeight != 0 && scaffSegType != 0xFF)
+                {
+                    const auto* scaffObj = ObjectManager::get<ScaffoldingObject>();
+                    const auto segmentHeight = scaffObj->segmentHeights[scaffSegType];
+                    const auto& scaffImages = getScaffoldingImages(scaffSegType).get2x2();
+                    uint32_t baseScaffImageIdx = scaffObj->image;
+                    ImageId baseScaffImage{};
+                    if (elIndustry.isGhost())
+                    {
+                        baseScaffImage = Gfx::applyGhostToImage(baseScaffImageIdx);
+                    }
+                    else
+                    {
+                        baseScaffImage = ImageId(baseScaffImageIdx, scaffoldingColour);
+                    }
+                    auto scaffImage = baseScaffImage.withIndexOffset(scaffImages.part1);
+                    height = baseHeight;
+                    for (int8_t remainingHeight = totalSectionHeight; remainingHeight > 0; remainingHeight -= segmentHeight, height += segmentHeight)
+                    {
+                        session.addToPlotListAsChild(scaffImage, { 0, 0, height }, { -8, -8, bbZOffset }, { 38, 38, bbLengthZ });
+                    }
+                    scaffImage = baseScaffImage.withIndexOffset(scaffImages.getRoof(rotation));
+                    session.addToPlotListAsChild(scaffImage, { 0, 0, height }, { -8, -8, bbZOffset }, { 38, 38, bbLengthZ });
+                }
+            }
+            session.setSegmentSupportHeight(Segment::all, 0xFFFF, 0);
+            session.setGeneralSupportHeight(0xFFFF, 0); // TODO: Check if this works previously would not set slope to zero
+        }
+        else
```

Comment:
This is a whole load of duplication which can probably be removed with a templated function.

Reply:
even without templating, you can reduce some duplicate code with simple shared method. Replace your PaintIndustry.cpp with this:

```
#include "PaintIndustry.h"
#include "../Industry.h"
#include "../Map/Tile.h"
#include "../Objects/IndustryObject.h"
#include "../Objects/ObjectManager.h"
#include "../Objects/ScaffoldingObject.h"
#include "../ScenarioManager.h"
#include "../Ui.h"
#include "Paint.h"

namespace OpenLoco::Paint
{
    void paintIndustryCore(PaintSession& session, const Map::IndustryElement& elIndustry, uint8_t totalSectionHeight, uint8_t numSections, uint8_t sectionProgress, uint32_t ticks, stdx::span<const std::uint8_t>& animationSequence, ImageId baseColour, Map::Pos3 offset, Map::Pos3 bbOffset, Map::Pos3 bbLength)
    {
        auto* industry = elIndustry.industry();
        auto* indObj = industry->getObject();
        const auto baseHeight = elIndustry.baseHeight();
        // 0x00525D30 (note this should be combined with a applyGhostToImage)
        const auto scaffoldingColour = indObj->scaffoldingColour;
        uint32_t buildingType = elIndustry.buildingType();
        const auto buildingParts = indObj->getBuildingParts(buildingType);
        // Combine this with any imageId
        const uint8_t rotation = (session.getRotation() + elIndustry.rotation()) & 0x3;

        const auto scaffSegType = indObj->scaffoldingSegmentType;
        if (totalSectionHeight != 0 && scaffSegType != 0xFF)
        {
            const auto* scaffObj = ObjectManager::get<ScaffoldingObject>();
            const auto segmentHeight = scaffObj->segmentHeights[scaffSegType];
            const auto& scaffImages = getScaffoldingImages(scaffSegType).get1x1();
            uint32_t scaffImageIdx = scaffObj->image + scaffImages.part0;
            ImageId scaffImage{};
            if (elIndustry.isGhost())
            {
                scaffImage = Gfx::applyGhostToImage(scaffImageIdx);
            }
            else
            {
                scaffImage = ImageId(scaffImageIdx, scaffoldingColour);
            }
            auto height = baseHeight;
            for (int8_t remainingHeight = totalSectionHeight; remainingHeight > 0; remainingHeight -= segmentHeight, height += segmentHeight)
            {
                session.addToPlotListAsChild(scaffImage, { offset.x, offset.y, height }, bbOffset, bbLength);
            }
        }

        int8_t sectionCount = numSections;
        auto height = baseHeight;

        for (const auto buildingPart : buildingParts)
        {
            if (sectionCount == -1)
            {
                break;
            }
            auto& buildingAnimation = indObj->buildingPartAnimations[buildingPart];
            auto adjustedBuildingPart = buildingPart;

            if (buildingAnimation.numFrames)
            {
                auto frameMask = buildingAnimation.numFrames - 1;
                auto cl = buildingAnimation.animationSpeed & 0x7F;
                auto tickThing = ticks >> cl;
                if (buildingAnimation.animationSpeed & (1 << 7))
                {
                    auto pos = Map::TilePos2(session.getUnkPosition());
                    tickThing += pos.x * 5;
                    tickThing += pos.y * 3;
                }
                adjustedBuildingPart += frameMask & tickThing;
            }
            else
            {
                if (!animationSequence.empty())
                {
                    auto tickThing = (ticks >> buildingAnimation.animationSpeed) & (animationSequence.size() - 1);
                    adjustedBuildingPart += animationSequence[tickThing];
                }
            }
            const auto sectionHeight = indObj->buildingPartHeight[adjustedBuildingPart];
            const uint32_t imageIdx = adjustedBuildingPart * 4 + indObj->var_12 + rotation;
            ImageId image = baseColour.withIndex(imageIdx);
            if (sectionCount == 0 && !baseColour.isBlended())
            {
                image = image.withNoiseMask((sectionProgress + 1) & 0x7);
            }
            session.addToPlotListAsChild(image, { offset.x, offset.y, height }, bbOffset, bbLength);
            height += sectionHeight;
            sectionCount--;
        }

        if (totalSectionHeight != 0 && scaffSegType != 0xFF)
        {
            const auto* scaffObj = ObjectManager::get<ScaffoldingObject>();
            const auto segmentHeight = scaffObj->segmentHeights[scaffSegType];
            const auto& scaffImages = getScaffoldingImages(scaffSegType).get1x1();
            uint32_t baseScaffImageIdx = scaffObj->image;
            ImageId baseScaffImage{};
            if (elIndustry.isGhost())
            {
                baseScaffImage = Gfx::applyGhostToImage(baseScaffImageIdx);
            }
            else
            {
                baseScaffImage = ImageId(baseScaffImageIdx, scaffoldingColour);
            }
            auto scaffImage = baseScaffImage.withIndexOffset(scaffImages.part1);
            height = baseHeight;
            for (int8_t remainingHeight = totalSectionHeight; remainingHeight > 0; remainingHeight -= segmentHeight, height += segmentHeight)
            {
                session.addToPlotListAsChild(scaffImage, { offset.x, offset.y, height }, bbOffset, bbLength);
            }
            scaffImage = baseScaffImage.withIndexOffset(scaffImages.getRoof(rotation));
            session.addToPlotListAsChild(scaffImage, { offset.x, offset.y, height }, bbOffset, bbLength);
        }
    }

    // 0x00453C52
    void paintIndustry(PaintSession& session, const Map::IndustryElement& elIndustry)
    {
        session.setItemType(Ui::ViewportInteraction::InteractionItem::industry);
        auto* industry = elIndustry.industry();
        auto* indObj = industry->getObject();

        ImageId baseColour(0, elIndustry.var_6_F800());
        if (elIndustry.isGhost())
        {
            baseColour = Gfx::applyGhostToImage(0);
        }
        // Combine this with any imageId
        const uint8_t rotation = (session.getRotation() + elIndustry.rotation()) & 0x3;
        // 0xE0C3A0
        auto ticks = ScenarioManager::getScenarioTicks();
        uint8_t numSections = 0xF0; // 0xF0 represents all sections completed
        // Only used when under construction
        uint8_t sectionProgress = 0;
        if (!elIndustry.isConstructed())
        {
            ticks = 0;
            numSections = elIndustry.var_6_003F();
            sectionProgress = elIndustry.sectionProgress();
        }

        // 0x00525D5C
        uint8_t sequenceIndex = elIndustry.sequenceIndex();

        // 0x00525D4E
        const int16_t bbLengthZ = std::min(elIndustry.clearHeight() - elIndustry.baseHeight(), 128) - 2;

        // 0x00E0C3B0
        stdx::span<const std::uint8_t> animationSequence{};
        if ((elIndustry.var_6_003F() & (1 << 5)) && (elIndustry.var_6_003F() & (1 << 4)))
        {
            animationSequence = indObj->getAnimationSequence(elIndustry.var_6_003F() & 0x3);
        }

        // 0x00E0C3A4
        uint32_t buildingType = elIndustry.buildingType();
        const auto buildingParts = indObj->getBuildingParts(buildingType);
        const auto baseHeight = elIndustry.baseHeight();
        int16_t bbZOffset = baseHeight;
        bool isMultiTile = indObj->buildingSizeFlags & (1 << buildingType);

        session.resetLastPS(); // Odd...
        if (indObj->flags & IndustryObjectFlags::hasShadows)
        {
            if (session.getRenderTarget()->zoomLevel <= 1)
            {
                const auto shadowImageOffset = buildingType * 4 + indObj->var_0E + rotation;
                const ImageId shadowImage = baseColour.withIndex(shadowImageOffset).withTranslucency(Colours::getShadow(elIndustry.var_6_F800()));
                if (isMultiTile)
                {
                    session.addToPlotListAsChild(shadowImage, { 0, 0, baseHeight }, { -8, -8, bbZOffset }, { 38, 38, bbLengthZ });
                }
                else
                {
                    session.addToPlotListAsChild(shadowImage, { 16, 16, baseHeight }, { 3, 3, bbZOffset }, { 26, 26, bbLengthZ });
                }
            }
        }

        // 0x00525D4F
        uint8_t totalSectionHeight = 0;
        if (numSections != 0xF0)
        {
            int8_t sectionCount = numSections;
            for (const auto buildingPart : buildingParts)
            {
                totalSectionHeight += indObj->buildingPartHeight[buildingPart];
                sectionCount--;
                if (sectionCount == -1)
                {
                    totalSectionHeight = std::max<uint8_t>(1, totalSectionHeight);
                    break;
                }
            }
        }

        if (isMultiTile)
        {
            // 0x004540A6
            if ((sequenceIndex ^ (1 << 1)) == ((-session.getRotation()) & 0x3))
            {
                paintIndustryCore(session, elIndustry, totalSectionHeight, numSections, sectionProgress, ticks, animationSequence, baseColour, Map::Pos3{ 0, 0, 0 }, Map::Pos3{ -8, -8, bbZOffset }, Map::Pos3{ 38, 38, bbLengthZ });
            }

            session.setSegmentSupportHeight(Segment::all, 0xFFFF, 0);
            session.setGeneralSupportHeight(0xFFFF, 0); // TODO: Check if this works previously would not set slope to zero
        }
        else
        {
            paintIndustryCore(session, elIndustry, totalSectionHeight, numSections, sectionProgress, ticks, animationSequence, baseColour, Map::Pos3{ 16, 16, 0 }, Map::Pos3{ 3, 3, bbZOffset }, Map::Pos3{ 26, 26, bbLengthZ });
        }
    }
}

```

Reply:
I did something similar to this in the end